### ISSUE TITLE
Token types from internal rules end up in parser

### ISSUE THREAD

**Issue by andreasabel** [BNFC/bnfc#302]
In a grammar like
```
Parsable.  Exp ::= Integer ;
Internal.  Exp ::= String;
```
The definition of `String` should not end up in the lexer, and no parser for `String` should be generated.

**Comment by fonfalleh** [comment]
Hi,
I've been playing around writing a parser for Lilypond and I've been having an issue that I think relates to this issue.
In particular, the following should be equivalent, but I could only parse the latter: `c_\ff`, `c _\ff`.
I believe parsing fails because `c_` is a valid Ident, however I am not using any Idents in my bnfc rules.

I managed to work around this by removing removing the references to identifier character, or `$i` in (Haskell backend) the generated Lex.x file.

Will this be solved by this issue, or could it be a separate issue?
Kind regards, 
Jacob

**Comment by andreasabel** [comment]
Hello @fonfalleh, thanks for he report.

Atm BNFC does not correctly support grammars that do not have a notion of _identifier_.  This is because keywords are first parsed (more precisely: lexed) as identifiers, and then looked up in a keyword dictionary.  For instance, keyword "if" matches also the regex for an `Ident` and is lexed as that and then turned into a keyword token.

A fix would be to include keywords in the generated lexer definition as their separate rules.  However, this blows up the generated lexer significantly.  (This might even exceed the limits on the automaton size in e.g. in `ocamllex`.) Thus, in practice, one lexes keywords as identifiers.

**Comment by fonfalleh** [comment]
Thanks!
I make use of token for variable declarations and references, but they can only contain letters, so I didn't use Ident for that in particular.

Does this mean that everything that can be lexed must match an Ident? Not sure if I follow.

**Comment by andreasabel** [comment]
> Does this mean that everything that can be lexed must match an Ident? Not sure if I follow.

No, but BNFC puts a regex for `Ident` in the lexer definition, whether you use it or not.  This means that e.g. `c_\ff` will be lexed as `Ident("c_") \ff` rather than `c _\ff` because of the longest match rule (the match "c_" is longer than the match "c" and thus takes precedence).  This explains the problems you are experiencing with your generated lexer.

Your problem would be fixed if `Ident` was not included in the lexer definition, but then the lexer has handle keywords differently.

**Comment by fonfalleh** [comment]
I see, thank you!

**Comment by jasper-e** [comment]
> No, but BNFC puts a regex for `Ident` in the lexer definition, whether you use it or not. This means that e.g. `c_\ff` will be lexed as `Ident("c_") \ff` rather than `c _\ff` because of the longest match rule (the match "c_" is longer than the match "c" and thus takes precedence). This explains the problems you are experiencing with your generated lexer.
> 
> Your problem would be fixed if `Ident` was not included in the lexer definition, but then the lexer has handle keywords differently.

Let me ask for a command line option to omit auto-including Ident.
I do not want to have this as a default behavior, just would like to be able to have the flag in my Makefile when I need it...
Since long I used workarounds (diff/patch) to omit this but every now and then after an upgrade my patchfile gets broken...
